File: qqzeng-phone-6.0/java/PhoneSearch6Db.java
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public final class PhoneSearch6Db {
    private static final int HeaderSize = 32;
    private static final int PrefixCount = 200;
    private static final int BitmapPopCountOffset = 0x4E2;

    private byte[] data = new byte[0];
    private String[] regionIsps = new String[0];
    private final IndexEntry[] index = new IndexEntry[PrefixCount];

    private static class Holder {
        static final PhoneSearch6Db INSTANCE = new PhoneSearch6Db();
    }

    public static PhoneSearch6Db getInstance() {
        return Holder.INSTANCE;
    }

    private PhoneSearch6Db() {
        loadDatabase();
    }

    private void loadDatabase() {
        String filePath = Paths.get(System.getProperty("user.dir"), "qqzeng-phone-6.0.db").toString();
        try {
            data = Files.readAllBytes(Path.of(filePath));
            ByteBuffer buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);

            int[] header = new int[8];
            for (int i = 0; i < header.length; i++) {
                header[i] = buffer.getInt(i * 4);
            }

            int regionsStart = HeaderSize;
            int ispsStart = regionsStart + header[1];
            int indexStart = ispsStart + header[2];

            String regionsStr = new String(data, regionsStart, header[1], "UTF-8");
            String[] regions = regionsStr.split("&");
            String ispsStr = new String(data, ispsStart, header[2], "UTF-8");
            String[] isps = ispsStr.split("&");

            int entryOffset = header[3];
            regionIsps = new String[header[4]];
            for (int i = 0; i < regionIsps.length; i++) {
                int entry = (buffer.getShort(entryOffset + i * 2) & 0xFFFF);
                regionIsps[i] = regions[entry >> 5] + "|" + isps[entry & 0x1F];
            }

            int pos = indexStart;
            for (int i = 0; i < PrefixCount; i++) {
                if (pos + 12 > data.length) {
                    index[i] = new IndexEntry(0, 0);
                    pos += 12;
                    continue;
                }
                int prefix = buffer.getInt(pos);
                if (prefix == i) {
                    index[i] = new IndexEntry(
                            buffer.getInt(pos + 4),
                            buffer.getInt(pos + 8)
                    );
                    pos += 12;
                } else {
                    index[i] = new IndexEntry(0, 0);
                }
               
            }
        } catch (IOException e) {
            throw new RuntimeException("Database file not found", e);
        } catch (Exception e) {
            throw new RuntimeException("Invalid database format", e);
        }
    }

    public String query(String phone) {
        if (phone == null || phone.length() != 7 || !phone.matches("\\d{7}")) {
            throw new IllegalArgumentException("Invalid phone number");
        }

        int prefix = parsePhoneSegment(phone.substring(0, 3));
        int subNum = parsePhoneSegment(phone.substring(3, 7));

        if (prefix < 0 || prefix >= PrefixCount) {
            return null;
        }

        IndexEntry entry = index[prefix];
        if (entry.bitmapOffset == 0 || entry.dataOffset == 0) {
            return null;
        }

        int byteIndex = subNum >>> 3; // Equivalent to subNum / 8
        int bitIndex = subNum & 0x07; // Equivalent to subNum % 8

        if (entry.bitmapOffset + byteIndex >= data.length) {
            return null;
        }

        int bitmapValue = data[entry.bitmapOffset + byteIndex] & 0xFF;
        if ((bitmapValue & (1 << bitIndex)) == 0) {
            return null;
        }

        int popCountOffset = entry.bitmapOffset + BitmapPopCountOffset + (byteIndex << 1);
        if (popCountOffset + 2 > data.length) {
            return null;
        }
        int preCount = ByteBuffer.wrap(data, popCountOffset, 2)
                .order(ByteOrder.LITTLE_ENDIAN).getShort() & 0xFFFF;

        int localCount = Integer.bitCount(bitmapValue & ((1 << bitIndex) - 1));

        int dataPos = entry.dataOffset + (preCount + localCount) * 2;
        if (dataPos + 2 > data.length) {
            return null;
        }

        int entryIndex = ByteBuffer.wrap(data, dataPos, 2)
                .order(ByteOrder.LITTLE_ENDIAN).getShort() & 0xFFFF;

        return (entryIndex < regionIsps.length) ? regionIsps[entryIndex] : null;
    }

    private static int parsePhoneSegment(String segment) {
        int result = 0;
        for (char c : segment.toCharArray()) {
            if (c < '0' || c > '9') {
                throw new IllegalArgumentException("Invalid phone segment: " + segment);
            }
            result = result * 10 + (c - '0');
        }
        return result;
    }

    private static class IndexEntry {
        final int bitmapOffset;
        final int dataOffset;

        IndexEntry(int bitmapOffset, int dataOffset) {
            this.bitmapOffset = bitmapOffset;
            this.dataOffset = dataOffset;
        }
    }
}

File: qqzeng-phone-6.0/php/PhoneSearch6Db.php
<?php
 namespace mobile;
class PhoneSearch6Db {
    private const HeaderSize = 32;                // 8个uint的头部
    private const PrefixCount = 200;              // 电话号码前缀总数（0-199）
    private const BitmapPopCountOffset = 0x4E2;   // 位图统计信息偏移量

    private static $instance = null;
    private $data = '';
    private $regionIsps = [];
    private $index = [];

    private function __construct() {
        $this->loadDatabase();
    }

    public static function getInstance() {
        if (self::$instance === null) {
            self::$instance = new self();
        }
        return self::$instance;
    }

    private function loadDatabase() {
        $filePath = __DIR__ . '/qqzeng-phone-6.0.db';
        if (!file_exists($filePath)) {
            throw new Exception("Database file not found");
        }

        $this->data = file_get_contents($filePath);
        $offset = 0;

        // 解析头部（小端序）
        $header = [];
        for ($i = 0; $i < 8; $i++) {
            $header[] = unpack('V', substr($this->data, $offset, 4))[1];
            $offset += 4;
        }

        // 解析地区与运营商表
        $regionsStart = self::HeaderSize;
        $ispsStart = $regionsStart + $header[1];
        $indexStart = $ispsStart + $header[2];

        $regions = explode('&', substr($this->data, $regionsStart, $header[1]));
        $isps = explode('&', substr($this->data, $ispsStart, $header[2]));

        // 构建地区-运营商组合
        $entryOffset = $header[3];
        $this->regionIsps = [];
        for ($i = 0; $i < $header[4]; $i++) {
            $entry = unpack('v', substr($this->data, $entryOffset + $i * 2, 2))[1];
            $regionIndex = $entry >> 5;
            $ispIndex = $entry & 0x1F;
            $this->regionIsps[] = $regions[$regionIndex] . '|' . $isps[$ispIndex];
        }

        // 构建前缀索引表
        $offset = $indexStart;
        $this->index = array_fill(0, self::PrefixCount, ['bitmapOffset' => 0, 'dataOffset' => 0]);
        for ($i = 0; $i < self::PrefixCount; $i++) {
            $prefix = unpack('V', substr($this->data, $offset, 4))[1];
            if ($prefix === $i) {
                $bitmapOffset = unpack('V', substr($this->data, $offset + 4, 4))[1];
                $dataOffset = unpack('V', substr($this->data, $offset + 8, 4))[1];
                $this->index[$i] = [
                    'bitmapOffset' => $bitmapOffset,
                    'dataOffset' => $dataOffset
                ];
                $offset += 12;
            } else {
                $this->index[$i] = ['bitmapOffset' => 0, 'dataOffset' => 0];
            }
        }
    }

    public function query($phone) {
        if (strlen($phone) > 11 || !ctype_digit($phone)) {
            throw new InvalidArgumentException("Invalid phone number format");
        }

        $prefix = intval(substr($phone, 0, 3));
        $subNum = intval(substr($phone, 3, 4));

        if ($prefix < 0 || $prefix >= self::PrefixCount) {
            return null;
        }

        $indexEntry = $this->index[$prefix];
        if ($indexEntry['bitmapOffset'] === 0 || $indexEntry['dataOffset'] === 0) {
            return null;
        }

        $byteIndex = $subNum >> 3;
        $bitIndex = $subNum & 0b0111;

        if ($indexEntry['bitmapOffset'] + $byteIndex >= strlen($this->data)) {
            return null;
        }

        $bitmap = ord($this->data[$indexEntry['bitmapOffset'] + $byteIndex]);
        if (($bitmap & (1 << $bitIndex)) === 0) {
            return null;
        }

        $popCountOffset = $indexEntry['bitmapOffset'] + self::BitmapPopCountOffset + ($byteIndex << 1);
        $preCount = unpack('v', substr($this->data, $popCountOffset, 2))[1];

        $localCount = $this->countBits($bitmap & ((1 << $bitIndex) - 1));

        $dataPos = $indexEntry['dataOffset'] + (($preCount + $localCount) << 1);
        $entry = unpack('v', substr($this->data, $dataPos, 2))[1];

        return $this->regionIsps[$entry] ?? null;
    }

    private function countBits($num) {
        $count = 0;
        while ($num > 0) {
            $count += $num & 1;
            $num >>= 1;
        }
        return $count;
    }
}
?>